src/providers/oracle/qgsoracleprovider.cpp

```diff
@@ -3807,4 +3807,88 @@ QGISEXTERN QgsProviderGuiMetadata *providerGuiMetadataFactory()
 }
 #endif
 
+QVariantMap QgsOracleProviderMetadata::decodeUri( const QString &uri ) const
+{
+  const QgsDataSourceUri dsUri { uri };
+  QVariantMap uriParts;
+
+  if ( ! dsUri.database().isEmpty() )
+    uriParts[ QStringLiteral( "dbname" ) ] = dsUri.database();
+  if ( ! dsUri.host().isEmpty() )
+    uriParts[ QStringLiteral( "host" ) ] = dsUri.host();
+  if ( ! dsUri.port().isEmpty() )
+    uriParts[ QStringLiteral( "port" ) ] = dsUri.port();
+  if ( ! dsUri.service().isEmpty() )
+    uriParts[ QStringLiteral( "service" ) ] = dsUri.service();
+  if ( ! dsUri.username().isEmpty() )
+    uriParts[ QStringLiteral( "username" ) ] = dsUri.username();
+  if ( ! dsUri.password().isEmpty() )
+    uriParts[ QStringLiteral( "password" ) ] = dsUri.password();
+  if ( ! dsUri.authConfigId().isEmpty() )
+    uriParts[ QStringLiteral( "authcfg" ) ] = dsUri.authConfigId();
+  if ( dsUri.wkbType() != QgsWkbTypes::Type::Unknown )
+    uriParts[ QStringLiteral( "type" ) ] = dsUri.wkbType();
+  if ( uri.contains( QStringLiteral( "selectatid=" ), Qt::CaseSensitivity::CaseInsensitive ) )
+    uriParts[ QStringLiteral( "selectatid" ) ] = ! dsUri.selectAtIdDisabled();
+  if ( ! dsUri.table().isEmpty() )
+    uriParts[ QStringLiteral( "table" ) ] = dsUri.table();
+  if ( ! dsUri.schema().isEmpty() )
+    uriParts[ QStringLiteral( "schema" ) ] = dsUri.schema();
+  if ( ! dsUri.keyColumn().isEmpty() )
+    uriParts[ QStringLiteral( "key" ) ] = dsUri.keyColumn();
+  if ( ! dsUri.srid().isEmpty() )
+    uriParts[ QStringLiteral( "srid" ) ] = dsUri.srid();
+  if ( uri.contains( QStringLiteral( "estimatedmetadata=" ), Qt::CaseSensitivity::CaseInsensitive ) )
```

Comment:
Out of curiosity, do you know why we just don't use dsUri.useEstimatedMetadata().isEmpty() like other condition?
It look like we want to not care about case but in the end we still get dsUri.useEstimatedMetadata().

I noticed it's the same in postgres.

Reply:
@troopa81 I do not have a very clear conceptual view of the need of the encodeUri and decodeUri. But to me it looked like that we were actually 'marshalling' an uri to a QgsDataSourceUri (https://qgis.org/api/qgsdatasourceuri_8h.html) and viceversa. Because postgis kept all members of that class I though to let Oracle also keep all.
And I did not see the generic setParam yet... so the qgsdatasource can actually hold a whole set of custom params/key-value pairs.
I will add those, and I will comment the not used (just to be clear?). Thanks for checking!
Mmm probably I should also add that to postgis provider?

You question about "why we just don't use dsUri.useEstimatedMetadata().isEmpty()" I do not know, maybe has something todo with True vs true (python vs cpp?)? I tried git blame but could not find info about it.

Reply:
Mmm, creating an Oracle connection, and then looking into the project file to check the attributes of the Oracle (*connection*?) element I get the following items:
```
name="docker_localhost" 
host="localhost" 
port="1521" 
schema="" 
username="QGIS" 
password="qgis" 
saveUsername="true" 
savePassword="true" 
allowGeometrylessTables="true" 
userTablesOnly="true" 
database="XEPDB1" 
dbworkspace="" 
geometryColumnsOnly="true" 
estimatedMetadata="true" 
dboptions=""
```
If I create a layer from that connection and request the uri via python:
```
iface.mapCanvas().currentLayer().dataProvider().uri().uri()
```
I get (both in 3.16 and in current master)
```
'dbname=\'XEPDB1\' 
host=localhost 
port=1521 
user=\'QGIS\' 
password=\'qgis\' 
key=\'pk\' 
estimatedmetadata=true 
srid=4326 
type=Polygon 
table="QGIS"."SOME_POLY_DATA" (GEOM)'
```
I'm a little puzzled about what I can/should put in the QgsDataSourceUri now. I also see that QgsDataSourceUri itself as an encodedUri and setEncodedUri: https://github.com/qgis/QGIS/blob/master/src/core/qgsdatasourceuri.cpp#L610-L659
THAT one only returns the username password of the connection?

I removed the selectatid and sslmode, added dbworkspace and includegeoattributes (and checkPrimaryKeyUnicity to one side to make it symmetric)